### ISSUE TITLE
woocommerce_checkout_order_processed fires before order is actually processed

### ISSUE THREAD

**Issue by dsifford** [woocommerce/woocommerce#28113]
Please provide us with the information requested in this bug report. Without these details, we won't be able to fully evaluate this issue. 
Bug reports lacking detail, or for any other reason than to report a bug, may be closed without action.

**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**

> **Background Info:** We utilize our own transactional email/marketing provider so it was necessary to for us to disable all the WooCommerce-provided emails and in their place inject code that transacts with our transactional provider.

The action `woocommerce_checkout_order_processed` seems like it is positioned in the wrong place in the source code because it actually fires before the order is actually processed.

https://github.com/woocommerce/woocommerce/blob/537ec5d0e966ecb106eb3ab687fc41c2fce17561/includes/class-wc-checkout.php#L1144-L1162

This is not ideal for us because if we fire our transactional email off at that point then there is a possibility that order confirmation emails are sent when payment processing fails (mistyped CC info, etc).

We tried to circumvent this issue by instead hooking into the `woocommerce_thankyou` event, which does fire after the payment info processes, but it is subject to re-firing on page refresh, which is another no-go.

**Expected behavior**

Given the name of the action, I would assume that this action would fire when the order is _completely_ processed, rather than partially.

In other words, it would make the most sense to me if line 1156 (above) was moved below the payment processing steps (just below line 1162).

**Actual behavior**

The action fires before payment is actually processed.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**

1. Create an action that hooks into `woocommerce_checkout_order_processed`.
2. Checkout for the order, but type in incorrect CC info.
3. See that the action still fires.

**Screenshots**

N/A

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**

I can give you this if you need it, but it seems pretty cut and dry here.

(CC: @claudiosanches per our slack conversation)

**Comment by dsifford** [comment]
Update: Digging into this more I was able to determine that `woocommerce_payment_complete` is called even for non-payment orders. It appears that hooking into _this_ might resolve our particular issue, but I still stand by the original issue of `woocommerce_order_processed` would make far more sense if it fired after payment processing.

It should not have taken that much brain power to locate the other hook. Perhaps it's just me.

**Comment by DavidAnderson684** [comment]
"Given the name of the action, I would assume that this action would fire when the order is completely processed, rather than partially."

Well, another possibility is surely that the assumption is incorrect. You can just choose to look at it the other way round: the proper way to interpret a terse name which has potential ambiguity is via the context it's found in. Moving it into a different context, whilst keeping the name, and therein breaking every extension that uses it, just in order to maintain an assumption about what the naming ought to mean, doesn't sound like a good idea.

**Comment by dsifford** [comment]
Yep, the issue here is raised because I'm wondering what situation this could cause breaks if moved under payment processing. 

The action name is interpreted (by me) in a way that feels like it was originally intended to fire after the order was completely processed.

**Comment by dsifford** [comment]
That said, I didn't consider the possibility that an order can be considered "processed" even after being created (in a failed state). That case I suppose would require it to be in the position that it is in currently.

So feel free to close this issue if that is the case. Was just sort of unfortunate that it took me 3 tries to realize this and finally find that `woocommerce_payment_processed` hook.

**Comment by tobinfekkes** [comment]
@dsifford You are exactly right, this hook being before payment has caused many an issue for me too across various sites. 

Some sites make the transaction right on checkout (so you know immediately whether payment failed or not), and some make the transaction days later (when the item is actually shipped/delivered), and so we've had to adjust around this hook specifically (using payment_gateway hooks), as well as adding extra hooks later in the checkout flow to get their payment info/token/transaction_id off of it. This is also frustrating because not every site uses a payment gateway (maybe the store is for internal purposes), so abstracting this out across multiple sites and checkout flows to handle this particular hook situation has been....fun.

A hook on line 1163 after payment info is known would really work wonders and be clean, rather than digging around to follow the rabbit hole of hooks.

I've ended up using the ```woocommerce_payment_successful_result``` hook to do what I'd normally expect ```woocommerce_checkout_order_processed ```to do.  But I'm glad you were able to make ```woocommerce_payment_complete``` work for your scenario.

**Comment by DavidAnderson684** [comment]
> this hook being before payment has caused many an issue for me

I think you're both mixing up two separate things. The hook is where it is, useful for people who want one there (like me - I use it because I want, in at least one extension, to have a hook at the point that an order has finished checkout processing, which isn't the same as completing order processing). The other thing, which seems to be what you both want, is a different hook, at a later point. For that, I suggest you create a pull request.

**Comment by tobinfekkes** [comment]
That is correct. I'm not necessarily saying that hook is in the wrong spot, but more than since there is one before payment, it would be handy to have one after as well.

**Comment by megakraken** [comment]
so whatever became of this? I completely agree with @dsifford

**Comment by rrennick** [comment]
> The action fires before payment is actually processed.

@dsifford Thanks for creating the issue. I think what @DavidAnderson684 was referring to in his [first comment](https://github.com/woocommerce/woocommerce/issues/28113#issuecomment-717473879) is that the hook name is `woocommerce_checkout_order_processed` vs `woocommerce_checkout_payment_processed`.

In the scenario you used in the issue description (on-site payment method), the payment processing occurs within checkout order processing. However, with off-site payment methods like PayPal, it can take up to [4 days](https://developer.paypal.com/docs/api-basics/notifications/ipn/IPNIntro/) to complete the payment processing on an order.

I'll close the issue with a reference to [your alternate method](https://github.com/woocommerce/woocommerce/issues/28113#issuecomment-717479048) using the `woocommerce_payment_complete` hook.
